[2단계 - 기물의 확장, DB 적용] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
길이가 긴 메서드와 1 depth 이상인 코드가 있어서 선택했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
안녕하세요, 지노!

2단계에서는 첫번째로 환경적인 어려움에 봉착했는데, 검색을 통해 찾은 여러가지 방안을 시도해봤는데도 해결하지 못했습니다.. 😥

프로젝트의 docker 디렉토리에 docker-compose.yml 파일을 추가했는데, 해당 디렉토리에서 실행시킨 컨테이너를 사용하면 아래 오류가 발생하면서 MySQL 데이터베이스에 접근할 수 없습니다.

![image](https://github.com/user-attachments/assets/9532e76e-7fb8-46a3-97a4-5cfe3ce7226f)

`Host '...' is not allowed to connect` 오류 메세지를 해결하기 위해 docker-compose.yml 파일에 `MYSQL_ROOT_HOST` 값을 설정해보고 docker/db/mysql/init/init.sql 파일에서 수동으로 사용자를 추가하고 권한 부여도 해봤지만 이슈가 해결되지 않았습니다.. 해서 프로젝트가 아닌 외부에 따로 저장한 docker-compose.yml 파일로 Docker 컨테이너를 가동해 사용해야 했습니다:

1. docker-compose.yml 파일을 프로젝트 디렉토리가 아닌 외부에 따로 저장
2. 외부 디렉토리에서 Docker 컨테이너 실행: `docker-compose -p janggi up -d` → `docker exec -it $컨테이너_ID /bin/bash`
3. MySQL 접속: `mysql -u root -p` (비밀번호 root)
4. init.sql 파일의 쿼리문을 실행해서 데이터베이스 테이블 생성
5. 미션 Application 실행

인프라 사용이 처음이라 해결법을 찾는 데에 시간이 오래 걸려서 일단 위와 같이 처리했습니다 ㅠㅠ 환경 이슈는 계속 솔루션을 찾아보겠습니다!

코드 상으로는 아래와 같은 고민거리가 있습니다:

### 도메인 객체를 외부로 전달하는 방법
도메인에서 사용하는 기물의 타입, 팀, 좌표 정보를 DB에 저장하기 위해서는 Position과 Piece 객체의 상태값을 어떤 형태로든 DAO에게 전달해줘야 했습니다. 그 형태로는 DTO를 사용했는데, DTO에 담을 객체의 상태를 객체 자신이 직접 넣어서 DTO를 생성할지 혹은 상위 클래스에서 getter로 상태 값을 받아와 DTO를 생성할지를 고민하다가 getter를 사용하기로 했습니다. 객체가 자신이 가진 상태에 대한 DTO를 반환한다는 것은 외부에서 DTO를 필요로 한다는걸 알고 있다고 해석될 수 있어서였고, 두가지 방식을 비교해보았을 때 도메인 객체가 인프라의 요구사항을 모르게 하는 것이 더 우선순위가 높다고 판단돼서 getter를 사용하기로 결정했습니다.

DAO로 도메인 객체의 값을 전달하기 위해서는 위와 같은 상황이 불가피하다고 생각되는데, 혹시 이런 고민 과정에서 제가 놓치고 있는 포인트가 있을까요?

### 인프라 제어 메서드가 도메인으로 침투
장기 게임의 마지막 상태를 데이터베이스에 반영하기 위해, 기물이 한번 움직일 때마다 기물의 좌표를 DB에 업데이트하도록 구현했습니다. 그 과정에서 DB—도메인 사이의 컨트롤러 역할을 하는 JanggiData의 `update` 메서드를 내부 도메인인 Round에서 호출하게 되었는데, 이 부분이 인프라와의 소통을 위한 코드가 도메인에 침투한 것처럼 보여서 고민입니다.

기물이 이동함으로 인해 발생한 좌표 변동사항은 도메인 내부에서만 알 수 있기 때문에 인프라에 알리기 위해서는 위와 같이 콜백 형식으로 알려주는게 최선이라고 생각했습니다. 처음에는 옵저버 패턴을 고려해봤지만, 옵저버 또한 내부 도메인과 외부 인프라 객체를 함께 연결해준다는 문제는 동일하고 오히려 구조가 더 복잡해질 것 같아서 적용하지 않았습니다. 고민되는 코드는 코멘트로 표시해두겠습니다!

리뷰해주셔서 감사합니다!! 💗

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **새로운 기능**
  - 도커 기반 MySQL 데이터베이스 연동으로 게임 상태 저장 및 복구가 지원됩니다.
  - 게임 설명서에 장기 기물 이동 규칙, 게임판 구성 및 오류 메시지 등이 추가되어 보다 명확한 이용 가이드가 제공됩니다.

- **개선 사항**
  - 기물 이동 로직 및 점수 계산 체계가 개선되어 게임 플레이가 더욱 직관적으로 진행됩니다.
  - 플레이어 측면 및 기물 명칭의 컬러 표기가 개선되어 사용자 인터페이스가 한층 향상되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jbilee committed:
refactor(board): Board 메서드 호출 로직 수정

- 기물 이동 가능 여부 및 이동 기능을 호출하는 로직을 Board의 메서드로 이동

jbilee committed:
refactor(round): Round 클래스 생성

- 현재 차례와 Board를 상태로 저장하는 객체 생성

jbilee committed:
fix(round): 게임 종료 조건 로직 수정

- General 기물이 하나 이하가 남았을 때만 게임을 종료하도록 수정

jbilee committed:
refactor(piece): Piece에서 사용하는 Vector 값 상수화

jbilee committed:
feat(piece): 궁성 영역 및 궁성 이동 제한 구현

jbilee committed:
feat(piece): 기물 점수 구현

jbilee committed:
refactor(side): 도메인 클래스에서 뷰를 위한 코드 삭제

jbilee committed:
feat(db): 데이터베이스 연동

jbilee committed:
docs(readme): 기능 구현 목록에 데이터베이스 관련 내용 추가

jbilee committed:
chore: 패키지 수정 및 파일 이동

- movement 패키지 생성, 기물 이동에 사용되는 좌표, 벡터, 궁성 영역 클래스를 해당 패키지로 이동
- factory 패키지 삭제, PieceFactory 클래스를 대분류 역할을 하는 domain 패키지로 이동

jbilee committed:
feat(db): 마지막 차례 저장 기능 구현

jbilee committed:
chore: docker-compose.yml 및 init.sql 파일 추가

jbilee committed:
chore: init.sql 파일 이동

jbilee committed:
feat(db): 게임 데이터 삭제 기능 구현

- 플레이어 한쪽이 승리하면 기물 데이터를 삭제한 뒤에 애플리케이션이 종료되도록 구현

jbilee committed:
refactor(global): 클래스명 개선 및 메서드 분리

jbilee committed:
test(piece, board): 도메인 클래스에 대한 테스트 추가

jbilee committed:
test(dao): DAO 테스트 작성

jbilee committed:
style: 줄바꿈 컨벤션 일체화

jbilee committed:
refactor(position): PositionDTO를 Position 객체가 아닌 외부에서 생성하도록 변경

jbilee commented on src/main/java/janggi/domain/Round.java:
```diff
@@ -0,0 +1,49 @@
+        this.currentTurn = side;
+    }
+
+    public void commence(Position selectedPosition, Position targetPosition, BiConsumer<Position, Piece> dataConsumer) {
+        board.makeMove(currentTurn, selectedPosition, targetPosition);
+        if (board.hasBothGenerals()) {
+            changeTurn();
+        }
+        dataConsumer.accept(selectedPosition, board.getPiece(targetPosition));
+    }
```
외부(인프라) 메서드가 도메인으로 침투한 것으로 보이는 코드입니다. JanggiData 인스턴스를 직접적으로 전달해 janggiData.update를 호출하게 하는것보단 한단계 추상화시켜서 함수형 인터페이스를 사용했지만, 결국 Round가 dataConsumer의 존재를 아는 것은 동일한거 같습니다 😵

hyunssooo replied:
> 기물이 이동함으로 인해 발생한 좌표 변동사항은 도메인 내부에서만 알 수 있기 때문에 인프라에 알리기 위해서는 위와 같이 콜백 형식으로 알려주는게 최선이라고 생각했습니다. 

getter를 통해서 필요한 데이터를 저장할 수 있지 않을까요? 지금 updatePiece 메서드는 delete -> insert를 하고 있네요. update를 하지 않은 이유가 있을까요?

jbilee replied:
변경된 사항에 대한 결과값을 받아와서 외부 인프라에 전달해주는 것이기에 말씀주신 대로 getter로 전달하면 적절할 것 같습니다! 이 부분은 아래 커밋에서 적용해보았습니다!

[refactor(round): 도메인 클래스에서 데이터베이스 관련 메서드 호출 제거](https://github.com/woowacourse/java-janggi/pull/114/commits/35f578d49bc88842be7fde9599670f884ef664b3)

> 지금 updatePiece 메서드는 delete -> insert를 하고 있네요. update를 하지 않은 이유가 있을까요?

데이터베이스 동작에 대한 이해가 부족해서 primary key는 update할 수 없다고 착각했던 것 같습니다..! 기물이 움직이면 데이터베이스에서 해당 기물의 좌표만 targetPosition 값으로 update하도록 개선하였습니다. 추가로 상대 팀의 기물을 잡아도 잡힌 기물이 데이터베이스에서 삭제되지 않던 버그도 수정하였습니다.

[fix(dao): 기물 데이터 저장 관련 버그 수정](https://github.com/woowacourse/java-janggi/pull/114/commits/df167c8914ba071d8a46dcc9cb1dc675f3c571ee)
[refactor(dao): updatePiece 메서드 로직 개선](https://github.com/woowacourse/java-janggi/pull/114/commits/0b245c6f2032c828499b8115afd7681cc3fa5279)

hyunssooo commented on README.md:
````diff
@@ -96,3 +106,10 @@
     ```
     [안내] 한나라가 이겼습니다!
     ```
+
+## 데이터베이스 연동
+
````
간단하게 어플리케이션 실행 방법을 readme에 작성해 주시면 좋을 것 같아요. :)

jbilee replied:
피드백 감사합니다, README 문서 상단에 애플리케이션 실행 방법 추가하였습니다! 📃

[docs(readme): 애플리케이션 실행 방법 기재](https://github.com/woowacourse/java-janggi/pull/114/commits/42ddd8390f66fa2226dec8c904db5d41db8fb307)

hyunssooo commented on src/main/java/janggi/dao/JanggiDao.java:
```diff
@@ -0,0 +1,152 @@
+        return null;
+    }
+
+    public void addPieces(List<PieceDto> piece) {
+        String query = "INSERT INTO pieces(name, side, position_row, position_column) VALUES(?, ?, ?, ?)";
```
> 도메인 객체를 외부로 전달하는 방법
> 도메인에서 사용하는 기물의 타입, 팀, 좌표 정보를 DB에 저장하기 위해서는 Position과 Piece 객체의 상태값을 어떤 형태로든 DAO에게 전달해줘야 했습니다. 그 형태로는 DTO를 사용했는데, DTO에 담을 객체의 상태를 객체 자신이 직접 넣어서 DTO를 생성할지 혹은 상위 클래스에서 getter로 상태 값을 받아와 DTO를 생성할지를 고민하다가 getter를 사용하기로 했습니다. 객체가 자신이 가진 상태에 대한 DTO를 반환한다는 것은 외부에서 DTO를 필요로 한다는걸 알고 있다고 해석될 수 있어서였고, 두가지 방식을 비교해보았을 때 도메인 객체가 인프라의 요구사항을 모르게 하는 것이 더 우선순위가 높다고 판단돼서 getter를 사용하기로 결정했습니다.

> DAO로 도메인 객체의 값을 전달하기 위해서는 위와 같은 상황이 불가피하다고 생각되는데, 혹시 이런 고민 과정에서 제가 놓치고 있는 포인트가 있을까요?

네 저도 동일하게 생각합니다. getter를 사용하지 말라는 것 때문에 이런 고민을 하신 것 같은데요. getter를 사용하지 말라는건 비즈니스 로직을 처리할 때를 말합니다. 해당 경우에는 getter를 통해서 데이터를 꺼내야 한다고 저도 생각합니다. 모델에서 dto를 만들지 않은 것은 말씀처럼 모델이 dto를 모르는 것이고 이건 모델이 dto에 대한 의존성이 없을을 의미합니다. :)

jbilee replied:
피드백 감사합니다!! 덕분에 getter에 대한 제 기준도 더 확고하게 다질 수 있게 되었습니다 🤓

hyunssooo commented on src/main/java/janggi/manager/JanggiGame.java:
```diff
@@ -1,52 +1,42 @@
 
+    private final JanggiData janggiData;
     private final Viewer viewer;
+    private final Round round;
```
사용자가 여러명인 경우에 Round가 Controller의 필드에 위치해도 괜찮나요? 어떤 문제가 발생할 수 있을까요?

jbilee replied:
지금은 애플리케이션이 한 기기에서만 가동된다고 가정하고 있는데 만약 여러 사용자가 각자의 기기에서 애플리케이션을 실행한다고 하면 컨트롤러마다 Round가 따로 생성되어 애플리케이션 간의 Round 객체끼리 동기화되지 않는다는 문제가 있을 것 같습니다.

그런데 이 부분을 어떤 방향으로 개선하면 좋을지는 감이 오지 않습니다.. 공통적인 Round를 둔다고 하여도 애플리케이션끼리의 상호작용을 어떻게 구현해야 할지 잘 상상되지 않는 것 같습니다 😥 애플리케이션 A에서 접속하는 사용자의 차례가 끝났을 때 애플리케이션 B에서 접속하는 사용자의 터미널에 현재 장기판의 상태를 올바르게 출력해주고 싶다면 기술적으로 어떻게 접근하는게 좋을까요..?

hyunssooo replied:
요청마다 DB에서 해당 유저가 진행하고 있는 게임을 불러오면 되지 않을까요?

jbilee replied:
말씀하신 요청이 사용자가 콘솔에 입력하는 행동이라고 이해했는데, 그러면 사용자가 기물을 움직이는 행동을 수행할 때마다 데이터베이스에서 게임(현재 차례와 기물)에 대한 정보를 불러오도록 하는 것일까요? 지금은 Round 안에 board와 currentTurn이 상태로 있는데, 이 상태에 DB에서 불러온 값을 매번 새로 저장해주는 것을 말씀하신건지 궁금합니다! 😮

hyunssooo replied:
네 맞습니다! 콘솔 환경이 아닌 웹 환경이라고 하면 이해가 빠를 것 같아요!

jbilee replied:
확인 감사합니다, 조금씩 이해되는 것 같습니다! 이번 미션 기간 내에는 미처 적용해보지 못했지만 연습삼아 따로 구현하면서 학습해보겠습니다! 🤓

hyunssooo replied:
혹시라도 제가 필요하면 dm이나 코멘트 남겨주세요!

hyunssooo commented on docker/db/mysql/init/init.sql:
```diff
@@ -0,0 +1,28 @@
+	`last_moved` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+	PRIMARY KEY(`position_row`, `position_column`)
+);
+
```
PRIMARY KEY가 `position_row`, `position_column` 인데 piece가 움직이면 key 값이 계속 변하나요? 

현재 단일 사용자를 가정한 것 같은데요. 어플리케이션이 단일 사용자가 아닌 여러 사용자라면 어떤 문제점이 있을까요? 여러 게임이 동시에 진행 될 수 있다면 지금 테이블 구조로 가능할까요?

jbilee replied:
말씀하신대로 한번에 하나의 게임(물리적인 애플리케이션)에 하나의 사용자 혹은 두명의 사용자가 번갈아가면서 기기와 상호작용하는 것을 가정하고 기물의 좌표로 이루어진 primary key는 항상 하나뿐일 것이라 생각해 구현했는데, 여러 사용자나 여러 게임의 경우는 미처 고려해보지 못했습니다! 😳 그렇게 되면 기물의 좌표가 얼마든지 중복될 수 있어서, 테이블 구조를 수정해야 할 것 같습니다. 다수의 게임이 동시다발적으로 실행될 수 있다는 점을 감안해서 개선해보겠습니다!

jbilee replied:
Primary key를 기물에 좌표가 아닌 일반적인 id로 바꾸고, pieces 테이블에 game_id 열을 추가하여 어느 게임에 해당하는 piece인지를 알 수 있도록 하였습니다:

[feat: 게임 방 입장 기능 구현](https://github.com/woowacourse/java-janggi/pull/114/commits/4cf3faf74d24e2a2bfb7859d8ff67bf716ea4d05)

hyunssooo commented on src/main/java/janggi/domain/PieceName.java:
```diff
@@ -0,0 +1,66 @@
+
+    private final Piece piece;
+    private final String displayName;
+    private final String databaseName;
```
도메인 모델이 dbName까지 알아야 할까요?

jbilee replied:
도메인의 핵심 로직에서는 DB에서만 저장되는 값이 쓰일 일이 없어서, `databaseName`을 통해 불필요한 의존성을 만들었다고 생각됩니다. 온전히 DB가 도메인에서 사용되는 값을 받아와서 저장하는 구조가 유지되도록 DB에 Piece 데이터를 저장하는 방식을 바꾸고 `databaseName` 필드를 제거했습니다!

[refactor(piece): 도메인에서 데이터베이스 전용 값 제거](https://github.com/woowacourse/java-janggi/pull/114/commits/5c4f0598310444cc3ebbf68193e03ae6c5416d8a)

hyunssooo commented on src/main/java/janggi/domain/movement/Directions.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(directions);
+    }
```
그렇다고 하네요. ㅎㅎ record가 어떤걸 만들어 주는지 학습해 보면 좋을 것 같아요.

jbilee replied:
Record는 불변 객체를 생성하는 데에 필요한 boilerplate 코드를 줄여주는 데에 특화된 클래스며, 따라서 final 필드의 생성자와 equals/hashCode 메서드의 재정의를 기본적으로 제공한다는 특징을 가지고 있습니다. Record 클래스를 사용할 때 이 부분 유의하면서 코드를 작성하겠습니다! 🤓


[refactor(directions): 불필요한 equals/hashCode 재정의 제거](https://github.com/woowacourse/java-janggi/pull/114/commits/fdb3f2c167e598e9e5ede744a4f6a6e5befa897f)

hyunssooo commented on src/main/java/janggi/dao/DaoSettings.java:
```diff
@@ -0,0 +1,22 @@
+package janggi.dao;
+
+public class DaoSettings {
+
+    public final String server;
+    public final String database;
+    public final String username;
+    public final String password;
+    public final String option;
```
이 클래스를 dao가 의존하고 있는데요. 어떤 의도일까요? 그리고 이 클래스는 어떤 의도일까요?

jbilee replied:
프로덕션 vs 테스트 환경 등, 필요에 따라 별도의 Connection 설정을 반환할 수 있는 객체를 의도했었는데 단순히 문자열만 들고 있고 어떠한 검증이나 행동을 가지지 않는 무의미한 껍데기를 만든 것 같습니다 😥 대신 본 클래스에 로직을 추가해 아래 피드백 주신 Connection을 관리하는 클래스로 만들어볼 수 있을 것 같습니다! 👀

jbilee replied:
기본적으로 쓰이는 공통된 설정은 상수로 두고 데이터베이스 이름을 문자열로 받아 해당 데이터베이스에 연결된 Connection 객체를 반환하는 클래스로 수정해보았습니다:
[refactor(dao): Connection 생성 로직을 DAO 클래스에서 분리](https://github.com/woowacourse/java-janggi/pull/114/commits/58e4ca250e3d127e17dff9d422d1300118f6829f)

hyunssooo commented on src/main/java/janggi/dao/JanggiDao.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public PieceDto getLastMovedPiece() {
+        String query = "SELECT * FROM pieces ORDER BY last_moved DESC LIMIT 1";
+
+        try (Connection connection = getConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            ResultSet resultSet = preparedStatement.executeQuery();
+
+            if (resultSet.next()) {
+                return new PieceDto(
+                        resultSet.getString("name"),
+                        resultSet.getString("side"),
+                        resultSet.getInt("position_row"),
+                        resultSet.getInt("position_column")
+                );
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+
+        return null;
+    }
```
누구 차례인지 자체를 저장해도 괜찮지 않을까요?

jbilee replied:
처음에 차례를 따로 테이블에 저장하는 방식을 떠올리긴 했었는데, 테이블 하나에 데이터 row가 하나만 들어간다는 부분이 너무 어색하게만 보였습니다.... 아직 데이터베이스에 대한 저만의 기준이 없어서 데이터 한줄만 저장하기 위한 테이블을 따로 둬도 괜찮은지에 대한 판단이 잘 서지 않았습니다 😧

해서 마지막으로 기물을 움직인 시간을 저장하고 해당 기물의 팀을 토대로 다음 차례를 판단했던 건데, 위에 primary key와 관련해 피드백 주신 내용처럼 다수의 게임 상태를 저장하는 방향으로 데이터베이스 테이블을 수정하면 각 게임마다 차례를 column으로 둬서 저장해주면 될 것 같습니다!

jbilee replied:
데이터베이스에 각각의 게임을 저장하는 games 테이블을 생성하고 해당 테이블에 현재 누구의 차례인지를 저장하도록 개선해보았습니다!

[feat: 게임 방 입장 기능 구현](https://github.com/woowacourse/java-janggi/pull/114/commits/4cf3faf74d24e2a2bfb7859d8ff67bf716ea4d05)

hyunssooo commented on src/main/java/janggi/dao/JanggiDao.java:
```diff
@@ -0,0 +1,152 @@
+    public PieceDto getLastMovedPiece() {
+        String query = "SELECT * FROM pieces ORDER BY last_moved DESC LIMIT 1";
+
+        try (Connection connection = getConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(query)) {
```
이 부분은 계속 반복되는 것 같은데요. 중복을 제거해 볼 수 있지 않을까요?

jbilee replied:
try-with-resources 블럭을 재사용할 수 있는 메서드로 분리하고, 해당 메서드에는 상황에 따라 다른 각각의 PreparedStatement 활용 로직을 함수형 인터페이스로 전달하는 방식으로 중복으로 제거해보았습니다! 어떤 쿼리는 조회로 인한 반환 값이 있는 반면, 어떤 쿼리는 반환값이 없어서 시그니처를 통해 메소드를 오버로딩하였습니다.

[refactor(dao): 반복되는 로직을 메소드화하여 중복 제거](https://github.com/woowacourse/java-janggi/pull/114/commits/998334a78501a25b7a1b53ff5cdc005a5489d38f)

hyunssooo commented on src/main/java/janggi/dao/JanggiDao.java:
```diff
@@ -0,0 +1,152 @@
+            e.printStackTrace();
+            return null;
+        }
+    }
```
커넥션을 관리하는 클래스를 작성해 볼 수 있지 않을까요? DAO가 여러개라면 어떻게 관리할 수 있을까요?

jbilee replied:
DAO를 여럿 두어야 하는 경우 커넥션을 관리하는 클래스에 DriverManager API로 커넥션을 받아오는 통용적인 메서드를 만들어 상황에 따라 알맞는 데이터베이스 Connection을 반환하도록 구현해볼 수 있을 것 같습니다.

기존의 DaoSettings 객체를 DatabaseConnector라는 클래스로 재해석하여 실제로 Connection 객체를 리턴하도록 개선해보았습니다:
[refactor(dao): Connection 생성 로직을 DAO 클래스에서 분리](https://github.com/woowacourse/java-janggi/pull/114/commits/58e4ca250e3d127e17dff9d422d1300118f6829f)

hyunssooo commented on src/test/java/janggi/dao/JanggiDaoTest.java:
```diff
@@ -0,0 +1,97 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class JanggiDaoTest {
+
```
여기 테스트는 DB 때문에 다 실패하는데요. 지금은 db 관련 테스트를 작성하지 않으셔도 괜찮습니다. :) 
다만 추후에 DB가 필요한 테스트들은 어떻게 진행할 수 있을지 고민해 보셔도 좋습니다!

jbilee replied:
이번 미션의 [셀프체크](https://prolog.techcourse.co.kr/studylogs/4108) 문항에 대한 답변을 작성하면서 Fake 객체의 존재에 대해 알게 됐는데, 저는 제가 작성한 JDBC 쿼리와 관련된 로직을 테스트하고 싶은데 Fake 객체는 그런 용도로 쓰이지 않는 것으로 보여서 JanggiDaoTest에 적용하지 않았습니다.. 😅 DB와 관련된 테스트는 앞으로 DB와 JDBC를 더 심화적으로 학습하면서 고민해보려 합니다!

JanggiDaoTest는 데이터베이스가 연결되지 않은 환경에서는 반드시 실패하는 상태인데, 그렇다면 지금은 이 테스트 코드를 아예 삭제하는 것이 좋을까요? 😮

hyunssooo replied:
이번 미션에서는 삭제하셔도 괜찮을 것 같습니다. 여러가지 방법이 있는데 현재 도입하기에는 비용이 많이 들 것 같아서요. 
원하는 테스트를 위해서 테스트용 DB를 두는 것이 가장 나을 것 같아요. H2 혹은 Testcontainers 등이 있을 것 같아요.

jbilee replied:
의견과 정보 감사합니다! 🤓

DAO에 대한 테스트 코드는 삭제하였습니다!

[test(dao): JanggiDao 테스트 삭제](https://github.com/woowacourse/java-janggi/pull/114/commits/5a359da372bdc737718fec835640e08f22599960)

hyunssooo commented on src/test/java/janggi/domain/RoundTest.java:
```diff
@@ -0,0 +1,26 @@
+
+class RoundTest {
+
+    @DisplayName("한나라의 경우 기본 기물 점수에 덤으로 인한 추가 점수 1.5점을 더한다.")
+    @Test
+    void test1() {
```
오 이건 장기의 규칙인가요?

jbilee replied:
맞습니다, 초나라가 먼저 공격하기에 우위가 있어서 1.5점이라는 덤을 한나라에게 준다고 합니다!

> Because the player with the blue pieces (cho) starts, they have an advantage. To compensate for this, Red receives 1.5 points (called 'deom' (點; 덤) in Korean), the half-point being to avoid ties.

그런데 덤에 대한 내용이 기능 구현 목록에 기재가 안돼있는걸 깨달았습니다!! 이 규칙은 README에도 반영하겠습니다! 🙇‍♀️

jbilee replied:
[docs(readme): 누락된 2단계 기능구현 내용 추가](https://github.com/woowacourse/java-janggi/pull/114/commits/84a6c9023afb729251ceb0b121198cc911fa3a1f)

hyunssooo commented on src/main/java/janggi/domain/Round.java:
```diff
@@ -0,0 +1,49 @@
+import janggi.domain.piece.Piece;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+public class Round {
+
```
개인적인 생각인데.. round 보다 좋은 이름이 없으까요? 저번 step에서는 janggiGame 정도의 이름을 생각하긴 했거든요. 요 부분은 의견 주시면 좋을 것 같아요! 명확한 의도가 있으시면 설명만 주셔도 좋습니다!

jbilee replied:
새로운 장기 게임을 시작하고 승자가 결정되면 종료하는 하나의 사이클이라는 의미로 round를 사용했는데, 고민해보니 round 안에 동의어처럼 보일 수 있는 `currentTurn`이라는 변수명도 함께 위치해있어서 혼란을 부를 수 있을 것 같습니다. 저도 JanggiGame이 더 직관적이라고 생각돼서 Round는 JanggiGame으로, 현재의 JanggiGame은 GameController로 수정하는게 명확한 의미 전달 관점에서 더 좋을 것 같습니다! 😀

jbilee replied:
[refactor(global): 클래스명을 직관적으로 개선](https://github.com/woowacourse/java-janggi/pull/114/commits/4ebb8643cd947c57ae60571a1c5171a8b266e316)

hyunssooo reviewed (changes requested):
안녕하세요. 줄리! 
몇 가지 코멘트 남겼습니다! 
늦어서 죄송해요. 🥲

jbilee committed:
docs(readme): 애플리케이션 실행 방법 기재

jbilee committed:
docs(readme): 누락된 2단계 기능구현 내용 추가

jbilee committed:
refactor(directions): 불필요한 equals/hashCode 재정의 제거

jbilee committed:
refactor(piece): 도메인에서 데이터베이스 전용 값 제거

jbilee committed:
refactor(round): 도메인 클래스에서 데이터베이스 관련 메서드 호출 제거

jbilee committed:
refactor(dao): 반복되는 로직을 메소드화하여 중복 제거

jbilee committed:
refactor(global): 클래스명을 직관적으로 개선

- Round 클래스명을 JanggiGame으로 변경
- 기존 JanggiGame 클래스명을 GameController로 변경
- JanggiData 클래스명을 DataController로 변경

jbilee committed:
refactor(dao): Connection 생성 로직을 DAO 클래스에서 분리

jbilee committed:
fix(dao): 기물 데이터 저장 관련 버그 수정

- 잡힌 기물이 데이터베이스에서 삭제되지 않은 이슈 해결
- 움직인 기물을 삭제하는 대신 변경된 값으로 UPDATE하도록 수정

jbilee committed:
test(dao): JanggiDao 테스트 삭제

jbilee committed:
refactor(dao): updatePiece 메서드 로직 개선

- 데이터베이스 업데이트에 필요하지 않은 값 제외

jbilee committed:
feat: 게임 방 입장 기능 구현

- 데이터베이스에 games 테이블 추가

jbilee committed:
refactor: 미사용 메서드 삭제

jbilee committed:
fix: Direction 내 누락되었던 상수 추가

hyunssooo commented on README.md:
```diff
@@ -1,58 +1,105 @@
+# 장기 미션 애플리케이션 실행 방법
+
+본 애플리케이션은 MySQL 데이터베이스를 활용합니다. MySQL 데이터베이스에 연결하기 위해 아래 절차를 따라 Docker 컨테이너를 실행시킨 후에 애플리케이션을 사용해야 합니다.
+
+1. 프로젝트의 ./docker 디렉터리에서 Docker 컨테이너 실행: `docker-compose -p janggi up -d`
+2. 애플리케이션 실행
+
+참고: Docker 컨테이너를 실행할 땐 외부 터미널을 사용합니다. IDE 내부 터미널에서 실행할 경우 "Can't connect to local MySQL server through socket '
+/var/run/mysqld/mysqld.sock'" 오류가 발생할 수 있습니다.
+
+애플리케이션을 처음 사용할 경우 초기 데이터베이스 테이블을 생성해야 합니다.
+
+1. Docker 컨테이너 실행: `docker-compose -p janggi up -d`
+2. 컨테이너의 shell에 접근: `docker exec -it $컨테이너_ID /bin/bash` ($컨테이너_ID는 `docker ps`로 확인)
+3. MySQL에 접속: `mysql -u root -p` (비밀번호: root)
+4. `./docker/db/mysql/init/init.sql`에 기재된 쿼리문 실행
+
```
👍

hyunssooo reviewed (approved):
안녕하세요. 줄리! 
이번 미션 진행하신다고 고생 많으셨습니다! 
방학 잘 보내시고 레벨2도 화이팅이에요!
